binding/HarfBuzzSharp.Shared/Buffer.cs

```diff
@@ -44,15 +43,37 @@ public uint Length
 			get { return HarfBuzzApi.hb_buffer_get_length(Handle); }
 		}
 
-		public void AddUtf8(string utf8text)
+		public void AddUtf8(string text)
 		{
-			var bytes = Encoding.UTF8.GetBytes(utf8text);
+			var bytes = Encoding.UTF8.GetBytes(text);
 			AddUtf8(bytes);
 		}
 
-		public void AddUtf8(byte[] bytes)
+		public void AddUtf8(byte[] text, uint itemOffset = 0, int itemLength = -1)
 		{
-			HarfBuzzApi.hb_buffer_add_utf8(Handle, bytes, bytes.Length, 0, bytes.Length);
+			HarfBuzzApi.hb_buffer_add_utf8(Handle, text, text.Length, itemOffset, itemLength);
+		}
+
+		public void AddUtf16(string text)
+		{
+			var bytes = Encoding.Unicode.GetBytes(text);
```

Comment:
I don’t know C#, but does this return UTF-16 code units or 8-bit bytes? Because the HarfBuzz functions expects code units of the respective encoding (8-bit, 16-bit or 23-bit code units).

Reply:
This returns UTF-16 as bytes.

**https://unicode-table.com/de/0041/**
`Encoding.UTF8.GetBytes("A") = {65}`
`Encoding.Unicode.GetBytes("A") = {65, 0}`
`Encoding.UTF32.GetBytes("A") = {65, 0, 0, 0}`

**https://unicode-table.com/de/1F600/**
`Encoding.UTF8.GetBytes("😀") = {240, 159, 152, 128}`
`Encoding.Unicode.GetBytes("😀") = {61, 216, 0, 222}`
`Encoding.UTF32.GetBytes("😀") = {0, 246, 1, 0}`

The second example should show the difference. Strings in C# are UTF-16(Unicode) encoded

Reply:
I can’t make sense of the UTF-16 and UTF-32 numbers, HarfBuzz wants something like `{0xD83D, 0xDE00} ` and `{0x1F600}`, respectively (using hex here for clarity).

Reply:
That is exactly what it produces.

**Full example**:
```			
var bytes = Encoding.Unicode.GetBytes("😀"); = {240, 159, 152, 128}

var utf16 = new ushort[2]; 

Buffer.BlockCopy(bytes, 0, utf16,0, 4); { utf16 = { 55357, 56832 } }

var highSurrogate = "0x" + utf16[0].ToString("X"); = { 0xD83D }
var lowSurrogate = "0x" + utf16[1].ToString("X"); = { 0xDE00 }

bytes = Encoding.UTF32.GetBytes("😀");

var utf32 = new uint[1]; = { 128512 }

Buffer.BlockCopy(bytes, 0, utf32, 0, 4);

var codepoint = "0x" + utf32[0].ToString("X"); = { 0x1F600 }
```